client/allocrunner/task_coordinator_test.go

```diff
@@ -0,0 +1,449 @@
+
+	// All tasks start blocked.
+	requireTaskBlocked(t, coord, init1Task)
+	requireTaskBlocked(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Set initial state, prestart tasks are allowed to run, mais is blocked.
+	states := map[string]*structs.TaskState{
+		init1Task.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		init2Task.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, init1Task)
+	requireTaskAllowed(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Init task dies, main is still blocked.
+	states = map[string]*structs.TaskState{
+		init1Task.Name: {
+			State:      structs.TaskStateDead,
+			Failed:     false,
+			StartedAt:  time.Now(),
+			FinishedAt: time.Now(),
+		},
+		init2Task.Name: {
+			State:     structs.TaskStateDead,
+			Failed:    true,
+			StartedAt: time.Now(),
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, init1Task)
+	requireTaskAllowed(t, coord, init2Task)
+	requireTaskBlocked(t, coord, mainTask)
+}
+
+func TestTaskCoordinator_SidecarNeverStarts(t *testing.T) {
+	ci.Parallel(t)
+
+	logger := testlog.HCLogger(t)
+
+	alloc := mock.LifecycleAlloc()
+	tasks := alloc.Job.TaskGroups[0].Tasks
+
+	mainTask := tasks[0]
+	sideTask := tasks[1]
+	initTask := tasks[2]
+
+	shutdownCh := make(chan struct{})
+	defer close(shutdownCh)
+	coord := newTaskCoordinator(logger, tasks, shutdownCh)
+
+	// All tasks start blocked.
+	requireTaskBlocked(t, coord, initTask)
+	requireTaskBlocked(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Set initial state, prestart tasks are allowed to run, mais is blocked.
+	states := map[string]*structs.TaskState{
+		initTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		sideTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, initTask)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+
+	// Init completes, but sidecar not yet.
+	states = map[string]*structs.TaskState{
+		initTask.Name: {
+			State:      structs.TaskStateDead,
+			Failed:     false,
+			StartedAt:  time.Now(),
+			FinishedAt: time.Now(),
+		},
+		sideTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+		mainTask.Name: {
+			State:  structs.TaskStatePending,
+			Failed: false,
+		},
+	}
+	coord.taskStateUpdated(states)
+	requireTaskAllowed(t, coord, initTask)
+	requireTaskAllowed(t, coord, sideTask)
+	requireTaskBlocked(t, coord, mainTask)
+}
+
+func TestTaskCoordinator_PoststartStartsAfterMain(t *testing.T) {
```

Comment:
These tests are great: can they be meaningfully adapted to exercise the expected state transitions in the case of a task restart (of all the various lifecycles), or do you think that's out of scope for these tests?

Reply:
I've been trying to think how to incorporate restarts here (along with #10785), but I couldn't find a good way yet. Since the next step will be to fix the restart flow I think I will postpone additional tests until then.

Reply:
Makes sense.